_posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd

```diff
@@ -0,0 +1,326 @@
+other hand -- we'll subsume them all here under PixelCNN -- are generally known for their good results. They seem to involve
+some more alchemy[^1] though. Under those circumstances, what could be more welcome than an easy way of experimenting with
+them? Through TensorFlow Probability (TFP) and its R wrapper, [tfprobability](https://github.io/tfprobability), we now have
+such a way.
+
+[^1]: Alluding to Ali Rahimi's (in-)famous "deep learning is alchemy" [talk](https://www.youtube.com/watch?v=Qi1Yry33TQE) at
+    NeurIPS 2017. I would suspect that to some degree, that statement resonates with many DL practitioners -- although one
+    need not agree that more mathematical rigor is the solution.
+
+This post first gives an introduction to PixelCNN, concentrating on high-level concepts (leaving the details to the curious,
+to look them up in the respective papers). We'll then show an example of using `tfprobability` to experiment with the TFP
+implementation.
+
+## PixelCNN principles
+
+### Autoregressivity, or: We need (some) order
+
+The basic idea in PixelCNN is autoregressivity. Each pixel is modeled as depending on all prior pixels. Formally:
+
+$$p(\mathbf{x}) = \prod_{i}p(x_i|x_0, x_1, ..., x_{i-1})$$
+
+Now wait a second - what even *are* prior pixels? Last I saw one images were two-dimensional. So this means we have to impose
+an *order* on the pixels. Commonly this will be *raster scan* order: row after row, from left to right. But when dealing with
+color images, there's something else: At each position, we actually have *three* intensity values, one for each of red, green,
+and blue. The original PixelCNN paper[@OordKK16] carried through autoregressivity here as well, with a pixel's intensity for
+red depending on just prior pixels, those for green depending on these same prior pixels but additionally, the current value
+for red, and those for blue depending on the prior pixels as well as the current values for red and green.
+
+$$p(x_i|\mathbf{x}<i) = p(x_{i,R}|\mathbf{x}<i)\ p(x_{i,G}|\mathbf{x}<i, x_{i,R})\ p(x_{i,B}|\mathbf{x}<i, x_{i,R}, x_{i,G})$$
+
+Here, the variant implemented in TFP, PixelCNN++[@Salimans2017PixeCNN] , introduces a simplification; it factorizes the joint
+distribution in a less compute-intensive way.[^2]
+
+[^2]: For details, see [@Salimans2017PixeCNN].
+
+Technically, then, we know how autoregressivity is realized; intuitively, it may still seem surprising that imposing a raster
+scan order "just works" (to me, at least, it is). Maybe this is one of those points where compute power successfully
+compensates for lack of an equivalent of a cognitive prior.
+
+### Masking, or: Where not to look
+
+Now, PixelCNN ends in "CNN" for a reason -- as usual in image processing, convolutional layers (or blocks thereof) are
+involved. But -- is it not the very nature of a convolution that it computes an average of some sorts, looking, for each
```

Comment:
```suggestion
involved. But -- is it not the very nature of a convolution that it computes an average of some sorts, looking for each
```

Reply:
same here... I would think the comma is needed ... perhaps the word "position" is a bad choice here (and above)... Perhaps I should call this "for each convolution", or even, "for each convolution output" ...

Reply:
yeah, I think the replacement made it clearer now